### DISCUSSION TITLE
Trying to get an SSD1357z 64x64 display working

### DISCUSSION THREAD

**Discussion by kevinfulness** [Bodmer/TFT_eSPI#2839] [Q&A - General]
I got this little display for esp32 AR projects and it's mostly working via the SSD1351 driver, however half the screen is garbled.
I've tried different frequencies with no results. I'm wondering if my pins are wrong, or if there's anything else I can change in the 1351 driver to make it work for the 1357z.

Pins are as follows:
CS: 18 
RST: 19
SCLK: 21
MOSI: 22
DC: 23

**Comment by staberas** [comment]
I just got the same display as you, id like to know if you figured it out .

**Comment by namuol** [comment]
I'm getting very similar results with the TFT_eSPI SSD1351 driver. Specified `TFT_WIDTH/HEIGHT` to be `64` as well. Appears that red and blue channels are swapped also, FWIW.

Currently looking for a datasheet to help me understand what's going on, but the [SparkFun 64x64 OLED Library](https://github.com/sparkfun/SparkFun_RGB_OLED_64x64_Arduino_Library) _does_ work for me, so first I want to try modifying the [TFT_eSPI SSD1351 initialization code](https://github.com/Bodmer/TFT_eSPI/blob/5793878d24161c1ed23ccb136f8564f332506d53/TFT_Drivers/SSD1351_Init.h) to match the [initialization code in the SparkFun library](https://github.com/sparkfun/SparkFun_RGB_OLED_64x64_Arduino_Library/blob/26ac43d4ccc9606dc6b8305d1b1375271173b2a5/src/SparkFun_RGB_OLED_64x64.cpp#L40-L42).

Most of the [command codes specified the main SSD1357 header file](https://github.com/sparkfun/SparkFun_RGB_OLED_64x64_Arduino_Library/blob/26ac43d4ccc9606dc6b8305d1b1375271173b2a5/src/SparkFun_SSD1357_OLED.h#L97-L145) of the SparkFun library seem to match the ones from the TFT_eSPI 1351 initialization code, which is promising.

**Reply by staberas** [reply]
You can try this one, made a while after searching and modifying around
https://github.com/staberas/SSD1357Z-Micropython-lib

**Reply by namuol** [reply]
Awesome, thanks! Did you find a datasheet by any chance?

**Reply by staberas** [reply]
The same datasheets as you, after many tries only the waveshare code worked for me so you can reverse engineer your way around it

**Reply by staberas** [reply]
@namuol any success?

**Reply by namuol** [reply]
I’ve put this on the back burner for some other stuff, but I’ll be sure to post back here if I make any progress!

**Reply by a-vidhawan** [reply]
Hi, I'm not too familiar with circuit python. Can I just flash this onto an Arduino? Also, does your code have support for images? I don't think I see a method for that.

**Comment by krstau** [comment]
Hi.

I have spent the last days trying to implement a driver for the SSD1357 (64x64), and I finally have made it work.
Please see my fork: [TFT_eSPI](https://github.com/krstau/TFT_eSPI)

For some reason, it does not like the 16-bit transfer in the setWindow()-function, so it has to be sent as two 8-bit transfers.
Do not know if this has any impact, maybe someone smarter than me can figure it out. Spent some time figuring this out...

I have used the initialization sequence from this datasheet (madctl has been modified -> 0x72):
[DM-OLED060-671 Datasheet](https://cdn.shopify.com/s/files/1/0264/7629/files/DM-OLED060-671p.pdf?v=1661130739)

The command list is here:
[SSD1357 Command List](https://support.newhavendisplay.com/hc/en-us/article_attachments/4414477721751)

What remains to be done is the rotation and "madctl" settings. I have not figured out the correct values yet,
but everything works like it should in rotation = 0, including the colors. Color-order can be changed with madctl (see datasheet)
I have taken inspiration from the SSD1333 pull request for the SSD1357_Rotation.h:
[SSD1333 Pull Request](https://github.com/Bodmer/TFT_eSPI/pull/3268/commits/e5b6c85da202bc20c91cd92e8072a69f971e8fd1)

Would probably also be smart to implement user settings for the 96x64 and other common resolutions.
I have tried to add some flexibility for now but it should be easy to do for the other displays.

The driver works for my purposes as it is now (have tested it with RP Pico), but if you want to figure out the remaining stuff please do.

**Reply by a-vidhawan** [reply]
Hey, I'm trying to use your library, but the board I have comes with 3-wire SPI. Will your library work out of the box? I bought my displays off of aliexpress i think you would have too? The pins i have are GND VCC SCL SDA RES DC CS.

**Reply by krstau** [reply]
Hi. The library should work out of the box, but you have to define your pins in the _User_Setups/Setup139_SSD1357_64.h_-file. Which microcontroller are you using? I have also bought my displays from Aliexpress, photo of the ones I bought below. Please let me know if you need any guidance.
![ssd1357](https://github.com/user-attachments/assets/4868f21b-fae8-44af-95b5-a353a71aeda9)

**Reply by a-vidhawan** [reply]
currently I'm trying to use [these ESP32S](https://www.amazon.com/dp/B0D8S22V7L?ref=ppx_yo2ov_dt_b_fed_asin_title&th=1) boards i got off of amazon, don't have experience with esp32 though just worked on Arduino and CircuitPython.
![image](https://github.com/user-attachments/assets/ed7e335e-87df-4eaf-a7f6-cce24b85ca9e)

Also, I'm using the exact same displays as you. Did you need any resistors while wiring this up to the esp? Your library should work with the esp32s too i think.

**Reply by krstau** [reply]
Ok. Yes the ESP32S is compatible, they should work just fine.

No resistors are needed, just use jumper wires directly.

For the ESP32 the pins are defined by default in the display user setup file as:

```
#define TFT_MOSI 23
#define TFT_SCLK 18
#define TFT_DC   2
#define TFT_RST  4
#define TFT_CS   15
```
So connections are:
ESP32 -> SSD1357 display

GND -> GND
3V3 -> VCC
GPIO23 -> SDA
GPIO18 -> SCK
GPIO2 -> DC
GPIO4 -> RES
GPIO15 -> CS

I assume you have the Arduino IDE set up with the ESP32 core? You are able to compile and upload to it (e.g blink sketch)? I also assume you have downloaded my fork instead of the original TFT_eSPI-repo and installed it as a library?

If so, try the _examples/Test and diagnostics/Colour_Test/Colour_Test.ino_ sketch to test the display. No other modifications are needed as my fork already has the SSD1357 driver selected as the user setup.

**Reply by a-vidhawan** [reply]
![WhatsApp Image 2025-04-29 at 01 46 20_99ecdb4c](https://github.com/user-attachments/assets/06f30894-6eed-4846-b6e4-a27e1b9b7e53)
![WhatsApp Image 2025-04-29 at 01 46 07_fd134c22](https://github.com/user-attachments/assets/29b43c86-eb05-43b4-a5af-ebf4dd8cae31)

Here's what I see when I try to do that. Also here is my wiring for the screen, it's as specified by you.

My final goal is to run a few of these screens in parallel, and displaying different stuff on them. Do you think that is possible without modifying your library right now? Could you maybe explain the flow of the library a little?

**Reply by krstau** [reply]
It is probably because it is trying to load the default setup in addition to the SSD1357. 

Open the _User_Setup_Select.h_-file in the Documents/Arduino/libraries/TFT_eSPI-folder. Comment the following line (put // in front) and save:

`#include <User_Setup.h>           // Default setup is root library folder`

Yes,  that should be possible. Connect all SDA and SCL lines together, but you need separate CS for each display if you want them to run something else at the same time.

I have only written the display driver, not anything in terms of the library, so I cannot give you any guidance there. It is very well documented, so it should not be a problem.

**Reply by a-vidhawan** [reply]
Any pointers on getting 64x64 rgb images displayed on it? I tried y=using the flash PNG example, adding my own bitmap in the '.h' file, changing buffer size to 128, but I keep seeing a white screen.

Also, for the multiple displays, where are your Chip Selects and stuff defined? So I can learn from that and implement the code for running 9 simultaneously
